build-scripts/graal.js

```diff
@@ -0,0 +1,211 @@
+const {
+  GRAAL_OS,
+  GRAAL_FOLDER,
+  GRAAL_VERSION,
+  GRAAL_URL
+} = require('./graal-env');
+const TEMP_PATH = path.resolve(__dirname, '../temp');
+
+process.on('unhandledRejection', error => {
+  console.error(error);
+  process.exit(1);
+});
+
+// The latest compiler is now compatible with a released version of graal
+const GRAAL_BUILD_FROM_SRC = !Boolean('COMPILER_NIGHTLY' in process.env);
+
+/**
+ * Simple wrapper for the NodeJS spawn command to use promises
+ *
+ * @param {string} command
+ * @param {Object=} spawnOpts
+ * @return {!Promise}
+ */
+function runCommand(command, spawnOpts = {}) {
+  process.stdout.write(`${command}\n`);
+  return new Promise((resolve, reject) => {
+    const commandParts = command.split(/\s+/);
+    const opts = Object.assign({}, {
+      stdio: 'inherit'
+    }, spawnOpts);
+    const externalProcess = spawn(commandParts[0], commandParts.slice(1), opts);
+    externalProcess.on('error', err => {
+      process.stderr.write(`${err.toString()}\n`);
+      reject(err);
+    });
+    externalProcess.on('close', exitCode => {
+      if (exitCode != 0) {
+        process.stderr.write(`non zero exit code ${exitCode}\n`);
+        process.exit(1);
+        reject();
+      }
+      resolve();
+    });
+  });
+}
+
+// Build graal from source
+if (!fs.existsSync(TEMP_PATH)) {
+  fs.mkdirSync(TEMP_PATH);
+}
+
+const NATIVE_IMAGE_BUILD_ARGS = [
+  '-H:+JNI',
+  '--no-server',
+  '-H:+ReportUnsupportedElementsAtRuntime',
+  '-H:IncludeResourceBundles=com.google.javascript.rhino.Messages',
+  '-H:IncludeResourceBundles=org.kohsuke.args4j.Messages',
+  '-H:IncludeResourceBundles=org.kohsuke.args4j.spi.Messages',
+  '-H:IncludeResourceBundles=com.google.javascript.jscomp.parsing.ParserConfig',
+  `-H:ReflectionConfigurationFiles=${path.resolve(__dirname, 'reflection-config.json')}`,
+  '-H:IncludeResources="(externs.zip)|(.*(js|txt))"',
+  '-jar',
+  path.resolve(process.cwd(), 'compiler.jar')
+];
+let buildSteps = Promise.resolve();
+if (GRAAL_BUILD_FROM_SRC) {
+  const GRAAL_SRC_BASE = path.resolve(TEMP_PATH, 'graal');
+  const MX_SRC_BASE = path.resolve(TEMP_PATH, 'mx');
+  const JVMCI_VERSION = 'jvmci-0.44';
+
+  // Clone the graal and mx repositories
+  buildSteps = buildSteps
+      .then(() => {
+        if (fs.existsSync(GRAAL_SRC_BASE)) {
+          return runCommand('git fetch', {cwd: GRAAL_SRC_BASE});
+        } else {
+          return runCommand('git clone https://github.com/oracle/graal.git', {cwd: TEMP_PATH});
+        }
+      })
+      .then(() => runCommand('git checkout vm-1.0.0-rc3', {cwd: GRAAL_SRC_BASE}))
+      .then(() => {
+        if (!fs.existsSync(MX_SRC_BASE)) {
+          return runCommand('git clone https://github.com/graalvm/mx.git', {cwd: TEMP_PATH});
+        }
+      });
+
+  // Download the JDK with JVMCI support
+  let JDK_URL;
+  let JDK_FOLDER;
+  let JDK_PATH;
+  let JDK8_UPDATE_VERSION;
+  if (process.platform === 'darwin') {
+    JDK8_UPDATE_VERSION = '181';
```

Comment:
Please add a comment saying why this is the right version.
If there's relevant external documentation that could help a future reader know it is time to change this, a link to that would be great.

Same for the case below, please.

Thanks.

Reply:
As long as you are ok with a comment that says "This is the version that works - newer versions don't"